### PR TITLE
krunvm: add support for darwin

### PR DESCRIPTION
**Pull request by jiegec**
###### Description of changes

This pr allows krunvm to build and run on darwin:

```shell
macbookair@macbookair /V/D/nixpkgs (krunvm-darwin)> ./result/bin/krunvm create debian --name debian
Resolving "debian" using unqualified-search registries (/nix/store/cmjp18dcg117hkf0qlv0sxlkbmjbgi01-krunvm-0.2.1/etc/containers/registries.conf)
Trying to pull quay.io/debian:latest...
Trying to pull docker.io/library/debian:latest...
Getting image source signatures
Copying blob 114ba63dd73a done  
Copying config 585393df05 done  
Writing manifest to image destination
Storing signatures
microVM created with name: debian
macbookair@macbookair /V/D/nixpkgs (krunvm-darwin)> ./result/bin/krunvm start debian
root@debian:/# cat /etc/issue
Debian GNU/Linux 11 \n \l

root@debian:/# 
```

Some commands are learned from krunvm homebrew formulae:

- https://github.com/slp/homebrew-krun/blob/master/Formula/buildah.rb
- https://github.com/slp/homebrew-krun/blob/master/Formula/krunvm.rb
- https://github.com/slp/homebrew-krun/blob/master/Formula/libkrun.rb
- https://github.com/slp/homebrew-krun/blob/master/Formula/libkrunfw.rb

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/virtualization/krunvm/default.nix [hunk 1]
```diff
@@ -22,17 +27,35 @@ stdenv.mkDerivation rec {
...
+    libiconv
+  ];
 
   makeFlags = [ "PREFIX=${placeholder "out"}" ];
 
+  postInstall = ''
+    mkdir -p $out/etc
+    install -D -m755 ${buildah-unwrapped.src}/docs/samples/registries.conf $out/etc/containers/registries.conf
+    install -D -m755 ${buildah-unwrapped.src}/tests/policy.json $out/etc/containers/policy.json
+  '';
```
Comment (NickCao) [inline pkgs/applications/virtualization/krunvm/default.nix:48]: These config files should reside in `/etc/containers`, not hard coded in packages.

File: pkgs/development/libraries/libkrunfw/default.nix [hunk 2]
```diff
@@ -44,6 +51,7 @@ stdenv.mkDerivation rec {
     homepage = "https://github.com/containers/libkrunfw";
     license = with licenses; [ lgpl2Only lgpl21Only ];
     maintainers = with maintainers; [ nickcao ];
-    platforms = [ "x86_64-linux" "aarch64-linux" ];
+    broken = stdenv.isDarwin && stdenv.isx86_64;
```
Comment (NickCao) [inline pkgs/development/libraries/libkrunfw/default.nix:54]: So this only works on `aarch64-darwin`?

**Comment by NickCao** [comment]
And in general, why use prebuilt binaries, it shouldn't be that hard to build a linux kernel on darwin with nixpkgs.

File: pkgs/development/libraries/libkrunfw/default.nix (hunk 2 above)
Comment (jiegec) [reply pkgs/development/libraries/libkrunfw/default.nix:54]: Yes @ https://github.com/containers/krunvm/blob/main/README.md:


Supported platforms:
Linux/KVM on x86_64.
Linux/KVM on AArch64.
macOS/Hypervisor.framework on ARM64.

File: pkgs/applications/virtualization/krunvm/default.nix (hunk 1 above)
Comment (jiegec) [reply pkgs/applications/virtualization/krunvm/default.nix:48]: This is due to the logic of `krunvm` to find configuration:

https://github.com/containers/krunvm/blob/ba9e585085cd02d0912d0aa254617e4299ade2b2/src/utils.rs#L29-L38

Maybe we can patch this code and move that configuration elsewhere.

File: pkgs/development/libraries/libkrunfw/default.nix (hunk 2 above)
Comment (NickCao) [reply pkgs/development/libraries/libkrunfw/default.nix:54]: Then you can just add aarch64-darwin to platforms, it looks cleaner this way.

File: pkgs/applications/virtualization/krunvm/default.nix (hunk 1 above)
Comment (NickCao) [reply pkgs/applications/virtualization/krunvm/default.nix:48]: Maybe open an issue upstream to ask about it.

**Comment by jiegec** [comment]
> And in general, why use prebuilt binaries, it shouldn't be that hard to build a linux kernel on darwin with nixpkgs.

It will require the user to create a case sensitive volume and build linux on top of that.

**Comment by zowoq** [comment]
I've cherry picked the `buildah` commit to https://github.com/NixOS/nixpkgs/pull/187064 and made minor changes.

File: pkgs/development/libraries/libkrunfw/default.nix (hunk 2 above)
Comment (jiegec) [reply pkgs/development/libraries/libkrunfw/default.nix]: Done.

**Comment by NickCao** [comment]
> It will require the user to create a case sensitive volume and build linux on top of that.
That's a strong justification, I'm fine with this now.

**Comment by jiegec** [comment]
> @jiegec Please rebase and drop the `buildah` commit from this PR.

Done

File: pkgs/applications/virtualization/krunvm/default.nix (hunk 1 above)
Comment (jiegec) [reply pkgs/applications/virtualization/krunvm/default.nix]: I have moved the files to `share/krunvm`.

**Pull request merged**